lib/ecmascript.ts

```diff
@@ -1176,60 +1196,89 @@ function MergeLargestUnitOption<T extends Temporal.DateTimeUnit>(
+      Required<T> extends Record<K, unknown>
+      ? T
+      : never
+    : never;
+
+type Prop<T, K> = T extends any ? (K extends keyof T ? T[K] : undefined) : never;
+
+// Resolve copies the keys and values of a given object type so that TS will
+// stop using type names in error messages / autocomplete. Generally, those
+// names can be more useful, but sometimes having the primitive object shape is
+// significantly easier to reason about (e.g. deeply-nested types).
+// Resolve is an identity function for function types.
+type Resolve<T> =
+  // Re-mapping doesn't work very well for functions, so exclude them
+  T extends (...args: never[]) => unknown
+    ? T
+    : // Re-map all the keys in T to the same value. This forces TS into no longer
+      // using type aliases, etc.
+      { [K in keyof T]: T[K] };
+
+type FieldObjectFromOwners<OwnerT, FieldKeys extends AnyTemporalKey> = Resolve<
+  {
+    -readonly // All keys in FieldKeys are required and don't include undefined.
+    [k in FieldKeys]: Exclude<Prop<OwnerT, k>, undefined>;
+  } & {
+    -readonly // All keys not in FieldKeys are optional, and include undefined
+    [k in Exclude<Keys<OwnerT>, FieldKeys>]?: Prop<OwnerT, k> | undefined;
+  }
+>;
+
 export function PrepareTemporalFields<
-  FieldKeys extends ResolvedFieldTypes<FieldSpec[number]>,
-  FieldSpec extends ReadonlyArray<Entry<AnyTemporalKey>>
->(bag: Partial<Record<FieldKeys, unknown>>, fields: FieldSpec): Required<Owner<Exclude<FieldKeys, undefined>>>;
-export function PrepareTemporalFields<
-  FieldKeys extends ResolvedFieldTypes<FieldSpec[number]>,
-  FieldSpec extends ReadonlyArray<Entry<AnyTemporalKey>>
->(
-  bag: Partial<Record<FieldKeys, unknown>>,
-  fields: FieldSpec,
-  completeness: 'partial',
-  opts?: FieldPrepareOptions
-): Owner<Exclude<FieldKeys, undefined>>;
-export function PrepareTemporalFields<
-  FieldKeys extends ResolvedFieldTypes<FieldSpec[number]>,
-  FieldSpec extends ReadonlyArray<Entry<AnyTemporalKey>>
->(
-  bag: Partial<Record<FieldKeys, unknown>>,
-  fields: FieldSpec,
-  completeness: 'complete',
-  opts?: FieldPrepareOptions
-): Required<Owner<Exclude<FieldKeys, undefined>>>;
-export function PrepareTemporalFields<
-  FieldKeys extends ResolvedFieldTypes<FieldSpec[number]>,
-  FieldSpec extends ReadonlyArray<Entry<AnyTemporalKey>>
+  FieldKeys extends AnyTemporalKey,
+  OwnerT extends Owner<FieldKeys>,
+  // Constrains the Required keys to be a subset of the given field keys
+  // This could have been written directly into the parameter type, but that
+  // causes an unintended effect where the required fields are added to the list
+  // of field keys, even if that key isn't present in 'fields'.
+  RequiredFieldKeys extends FieldKeys,
+  RequiredFields extends ReadonlyArray<RequiredFieldKeys> | FieldCompleteness,
+  ReturnT extends RequiredFields extends 'partial' ? Partial<OwnerT> : FieldObjectFromOwners<OwnerT, FieldKeys>
```

Comment:
This is new syntax to me: `T extends A extends B` - is that the same as `T extends A&B`?

Also, this (still?) looks like "return-only generics".  It sounds like you said this is intended to be inferred from usage.  The problem with inferred return-only-generics is that it allows requiring arbitrary constraints on the type that comes out of a function, which obviously is infeasible.  See this [toy example](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMQMBJDAHgBUA+ACgEoAueSgbgFgAoLrVDEGIihgEAERwF4Aby7x4AI1gBrBswBuOLMA6cAvly5g8AZwzxgE5uMkBeQsTIMdFggDpFMFfVbwA9L-gMAAsBEAByY3hUHHgAdygAT3tSDAZ4I2QIYHgABxxjYyx5CCSlaNjAmN5IZFB4KAVlNIBbYiCcYFcuoA) for how this can happen even accidentally, when the type is inferred rather than explicitly specified.  I'd argue that return-only-generics can be appropriate if the generic is nested in a template parameter of the returned type (e.g. `declare function emptyList<T>(): T[]`) but this is not what's happening here.

Stepping back, what is the intended usage pattern here?  You've got _a lot_ of parameters, with no defaults - are they all intended to be inferred, or specified directly?

It can be useful to pull out named intermediate types (like `ReturnT`) as generics, but I'd expect them to be assigned directly via a default (and marked `private` if only [that feature](https://github.com/microsoft/TypeScript/issues/26242#issuecomment-1230903819) would land, which it never will).  But I'm puzzled about what this is doing.  Writing this with `=` instead of `extends` doesn't necessarily make it safer from tampering (i.e. intentionally specifying something arbitrary) but it _does_ prevent accidental misuse by calling it with code that with back-infer a more specific type than this is prepared to return.

I dunno if you're interested in this, but it's actually possible to make arguments mostly-private by writing them as `declare function foo<NoExplicitArgsPastThisPoint extends typeof privateSymbol, ...>` ([example](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gBxgSwDcoMQBlATwFsAjJCALngFcU8BHJhNK2iAbgCwAKGGhIsBADMWYDHlTxJAHgBySAKIAPbBDxg8GAIIwA5mnghNJFMHMZy2EEkk58REhRp0ANPAAqAHwAFLAmjH4AlOEA2gC6gkJi0HCKMnIKJkp+vmpaOnoGxmYWViA2dg5OLriExGQ8dPAAvPD2js6utR4NEL4Aqs3+ccGh4VHwfQnCyOhYmoOSQQAsAEwRCTOY8OSDmZj4KCYAPihMNCAwwQDkmlfrwkA)), which prevents tampering with them, other than passing an explicit `never`.

Reply:
It's not new syntax, just very hard to read :| 

It should be 

ReturnT extends (RequiredFields extends 'partial' ? Partial<OwnerT> : FieldObjectFromOwners<OwnerT, FieldKeys>).

Thanks for the discussion r.e. generic bounds and return-only-generics. Following that advise, I've moved as much of the type-magic out of the generic param list to helper types, retaining only FieldKeys and RequiredFields as the two parameters that need inference.

Reply:
Steve and I discussed this offline and the new implementation LGTM.